src/cargo/core/workspace.rs

```diff
@@ -552,6 +552,23 @@ impl<'cfg> Workspace<'cfg> {
         })
     }
 
+    /// Reset "default build" behavior for non-virtual workspaces, so that a build in
+    /// the workspace root would build the root package, not the packages specified
+    /// in `workspace.default-members`.
+    pub fn ignore_default_members(&mut self) -> &mut Workspace<'cfg> {
+        // If we're building a virtual workspace, then there is no root package to build
+        if self.is_virtual() {
+            return self;
+        }
+        // If we're not building the root package, then the default members do not matter
+        if self.current_manifest != self.root_manifest() {
+            return self;
+        }
+
+        self.default_members = vec![self.current_manifest.clone()];
```

Comment:
I am concerned that this solution misuses how [Package Selection](https://doc.rust-lang.org/cargo/commands/cargo-build.html?highlight=default-members#package-selection) works within a non-virtual workspace with `default-features` set. This is because it works by relying on [this part](https://github.com/rust-lang/cargo/blob/9467f81fbf340b70f7ab7e3290146350133f3ea3/src/cargo/ops/cargo_compile.rs#L413-L420) of creating the build context. 

The Cargo book explains what I mean [here](https://doc.rust-lang.org/cargo/reference/workspaces.html#package-selection) by saying:
> The optional default-members key can be specified to set the members to operate on when in the workspace root and the package selection flags are not used

Given what the book says and what I linked to says in the build context it might be a better idea to modify the `CompileOptions` to act as though `-p` was passed in the command line. 

I could be wrong with my concerns so I would wait until someone from the [Cargo team](https://www.rust-lang.org/governance/teams/dev-tools#cargo) comments as well.

Reply:
> it might be a better idea to modify the CompileOptions to act as though -p was passed in the command line.

It absolutely is the better idea. This solution is for-sure technical debt until someone can figure that out.

The problem is that `CompileOptions` owns a `BuildConfig` which ultimately owns a `RefCell` containing `TcpListener` (inside `RustfixDiagnosticServer`), so it's a totally non-trivial thing to e.g. clone. (And I'm not confident enough in how Cargo works to know if there's any easy solution to this...)

If there's an easy solution someone can point me to, I'm happy to switch over to it, but I saw that `TcpListener` and got spooked that there might be hidden architectural assumptions of only one `BuildConfig` in a `cargo` invocation, or something like that...

Reply:
Having a `fn ignore_default_members` in `core::workspace.rs` looks weird to me, as it only serves for `cargo-install` and no one use it (and shouldn't). 

Reusing `-p` might be an option. The problem is that cargo-install contains ad-hoc solution that reads into a workspace but not use the info everywhere. And that might break things if we reuse `-p` option. I always feel a headache looking into these small but significant differences.